### ISSUE TITLE
Investigate from_padded implementations correctness

### ISSUE THREAD

**Issue by drisspg** [pytorch/pytorch#84082]
## Summary
A gradient test involving these ops created here: #84078 is succeeding for float64 on cuda but failing for float16 and float32 on cuda. The input sizes are small enough were floating point accumulation should not be effecting this tests.

## Env Details:
Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.22.3
Libc version: glibc-2.27

Python version: 3.9.12 (main, Jun  1 2022, 11:38:51)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-1069-aws-x86_64-with-glibc2.27
Is CUDA available: N/A
CUDA runtime version: 11.6.112
GPU models and configuration:
GPU 0: NVIDIA A100-SXM4-40GB
GPU 1: NVIDIA A100-SXM4-40GB
GPU 2: NVIDIA A100-SXM4-40GB
GPU 3: NVIDIA A100-SXM4-40GB
GPU 4: NVIDIA A100-SXM4-40GB
GPU 5: NVIDIA A100-SXM4-40GB
GPU 6: NVIDIA A100-SXM4-40GB
GPU 7: NVIDIA A100-SXM4-40GB

Nvidia driver version: 510.47.03
cuDNN version: Probably one of the following:
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.0.5
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn.so.8.1.1
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.1.1
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.1.1
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.1.1
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.1.1
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.1.1
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.1.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A

Versions of relevant libraries:
[pip3] mypy==0.960
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.6
[pip3] torchdynamo==1.13.0.dev0
[conda] mkl                       2022.0.1           h06a4308_117
[conda] mkl-include               2022.0.1           h06a4308_117
[conda] numpy                     1.21.6                   pypi_0    pypi
[conda] torchdynamo               1.13.0.dev0               dev_0    <develop>

cc @ezyang @gchanan @zou3519 @cpuhrsch @jbschlosser @bhosmer @mikaylagawarecki

**Comment by drisspg** [comment]
```
import torch

device='cuda'
dtype=torch.float32

a = torch.randn(5, 4, device=device, dtype=dtype, requires_grad=True)
b = torch.randn(5, 4, device=device, dtype=dtype, requires_grad=True)
x = torch.nested_tensor([a, b])

x.to_padded_tensor(0).sum().backward()

torch.testing.assert_close(a.grad, torch.ones(2, 4, device=device, dtype=dtype))
torch.testing.assert_close(b.grad, torch.ones(5, 4, device=device, dtype=dtype))
```

**Comment by drisspg** [comment]
when dtype = torch.float64 is everything works as expected. for both float 32 and float 16 the first element of A is correct and that is it

**Comment by drisspg** [comment]
Since to_padded appears to be working correctly and in the backward pass we are calling `from_padded()` my hunch is that the cuda implementation for from_padded() is not working correctly

**Comment by dagitses** [comment]
High priority for possible silent correctness issue.

**Comment by drisspg** [comment]
This issue is because due to broadcasting

**Comment by drisspg** [comment]
The backward formula for to_padded is: 
``` Yaml
- name: to_padded_tensor(Tensor self, float padding, int[]? output_size=None) -> Tensor
  self: at::_nested_from_padded(grad, self._nested_tensor_size())
  padding: non_differentiable
  ```
  In this case grad is non_contiguous and therefore the invariant that the underlying buffer be contiguous is broken.
  
  A fix is: 
  ```Yaml
- name: to_padded_tensor(Tensor self, float padding, int[]? output_size=None) -> Tensor
  self: at::_nested_from_padded(grad.contiguous(), self._nested_tensor_size())
  padding: non_differentiable
  ```
  will create PR now

**Comment by drisspg** [comment]
Fixed with: #84217
